### ISSUE TITLE
Hash ignores object equality

### ISSUE THREAD

**Issue by vais** [opal/opal#1013]
According to the docs (http://ruby-doc.org/core-2.2.0/Hash.html#class-Hash-label-Hash+Keys), in Ruby:

> Two objects refer to the same hash key when their hash value is identical and the two objects are eql? to each other.

So, let's try this in `pry`:
```
$ pry
[1] pry(main)> class V;def hash;0;end;def eql?(a);false;end;end
=> :eql?
[2] pry(main)> {V.new=>456, V.new=>123}
=> {#<V:0x007f902c153f88>=>456, #<V:0x007f902c153f60>=>123}
[3] pry(main)> class V;def hash;0;end;def eql?(a);true;end;end
=> :eql?
[4] pry(main)> {V.new=>456, V.new=>123}
=> {#<V:0x007f902c8d2868>=>123}
```
You can see that objects that have the same `hash` but are **not** `eql?` to each other, *both* show up in the hash, whereas objects that **are** `eql?` to each other are added to the hash only once (last one wins).

Now let's try this in Opal:
```
$ bundle exec opal-repl
>> class V;def hash;0;end;def eql?(a);false;end;end
=> "eql?"
>> {V.new=>456, V.new=>123}
=> {#<V:0x9a>=>123}
```
In Opal, regardless of `eql?`, last one wins, and the resulting hash has only one key.

**Comment by vais** [comment]
Which is covered in RubySpec and is currently in `filters/bugs/hash.rb`, so we'll get to it eventually :smile: 
`fails "Hash#[]= stores unequal keys that hash to the same value"`
```ruby
  it "stores unequal keys that hash to the same value" do
    h = new_hash
    k1 = ["x"]
    k2 = ["y"]
    # So they end up in the same bucket
    k1.should_receive(:hash).and_return(0)
    k2.should_receive(:hash).and_return(0)

    h.send(@method, k1, 1)
    h.send(@method, k2, 2)
    h.size.should == 2
  end
```

**Comment by vais** [comment]
@eregon sorry to bother you with another Opal issue - I'm working on `Hash`, and there are a few specs that I really want Opal to be able to run, but since they rely on Integer and Float not being `eql?` to each other, I cannot run them under Opal (i.e. there is no distinction between `1` and `1.0` whatsoever in Opal). In all of the three cases below, I feel `1` and `1.0` were chosen for convenience, it is not essential to the "spirit" of the cases under test, but purely incidental.

Do you have any suggestions how to re-write the following specs so that they still test what they are supposed to test, but do so without replying on `1` not being `eql?` to `1.0`?

https://github.com/ruby/rubyspec/blob/master/core/hash/element_reference_spec.rb#L67
https://github.com/ruby/rubyspec/blob/master/core/hash/shared/eql.rb#L114
https://github.com/ruby/rubyspec/blob/master/core/hash/invert_spec.rb#L16

cc @meh @elia

**Comment by eregon** [comment]
Indeed, 1 and 1.0 are just convenient objects that are `==` but not `eql?`.
Using mocks would be more appropriate for this like:
```ruby
a = mock()
a.should_not_receive(:==)
a.should_receive(:eql?).at_least(1)
```

If it feels to verbose you could have them in a helper (`HashSpecs.new_key` ?) in https://github.com/ruby/rubyspec/blob/master/core/hash/fixtures/classes.rb.

**Comment by vais** [comment]
Sweet, thanks, @eregon, this helps!

**Comment by vais** [comment]
Ok, @eregon, something is weird here, and I can't figure out what. I'm loosing faith in computing, help!

I was playing with this spec (https://github.com/ruby/rubyspec/blob/master/core/hash/element_reference_spec.rb#L68), trying to convert it to using mocks, but it was not behaving as I was expecting it to, so, in the end, here is what I came up with do demonstrate the problem:
```ruby
  it "compares keys with eql? semantics" do
    k1 = mock('k1')
    k1.should_not_receive(:==)
    k1.should_not_receive(:===)
    k1.should_not_receive(:eql?)
    k1.should_not_receive(:equal?)

    k2 = mock('k2')
    k2.should_not_receive(:==)
    k2.should_not_receive(:===)
    k2.should_not_receive(:eql?)
    k2.should_not_receive(:equal?)

    new_hash(k1 => "x")[k2].should == nil
  end
```
And it passes! WTF:
```
$ bundle exec mspec core/hash/element_reference_spec.rb 
ruby 2.2.2p95 (2015-04-13 revision 50295) [x86_64-darwin14]
[/ | ==================100%================== | 00:00:00]      0F      0E

Finished in 0.004494 seconds

1 file, 13 examples, 41 expectations, 0 failures, 0 errors, 0 tagged
```
(The `should_not_receive(:===)` part was added out of pure desperation, btw)

I confirmed that the issue is not with mspec by adding `k1.should_not_receive(:hash)` which makes it fail as expected. So, either the title of the example is wrong, or the world is broken, or maybe I just need some sleep :stuck_out_tongue_winking_eye: But I wanted to share this with you in the hope that you'll set me straight and show what I'm doing wrong here. How does Ruby compare these keys if not by any of the means above??

**Comment by elia** [comment]
Dumb question, if the hash of k1 and k2 doesn't match why should it compare them with == === or eql? ?

**Comment by eregon** [comment]
Exactly, as @elia said, if the `#hash`-es do not match, then there is no need to call `eql?`.
So in this case you might want to stub `#hash` to make sure they are the same.

**Comment by eregon** [comment]
@vais BTW, it seems the 3 specs below already specify this behavior via mocks.
Therefore I would either remove the spec above or tag it as unsupported for Opal.

This kind of specs using core objects properties tend to be useful to catch related problems and to show what a typical Ruby user might expect, so I think there are worth keeping.
Unfortunately I am not aware of other core types which are #== and not #eql?, so probably the best is to tag these in Opal and make sure there is a proper spec for the behavior with mocks or not relying on specific Numeric classes.

**Comment by elia** [comment]
@vais should this be closed as fixed by #1035?

**Comment by vais** [comment]
@elia yep:
```
$ bundle exec opal-repl
>> class V;def hash;0;end;def eql?(a);false;end;end
=> "eql?"
>> {V.new=>456, V.new=>123}
=> {#<V:0x9a>=>456, #<V:0x9c>=>123}
```
:joy_cat:
